### PR TITLE
Add support for sorting caret ('^') higher than base version

### PR DESCRIPTION
**Pull request by ignatenkobrain**
1.1^20160101 means 1.1 version (base) and patches which were applied at
that date on top of it.

* 1.1^201601 > 1.1
* 1.1^201601 < 1.1.1

Second case might be fulfilled with tilde separator, but it's not obvious
and doesn't fulfil first case.

Signed-off-by: Igor Gnatenko <i.gnatenko.brain@gmail.com>

### CODE REVIEW

File: lib/rpmvercmp.c [hunk 1]
```diff
@@ -45,6 +45,20 @@ int rpmvercmp(const char * a, const char * b)
 	    continue;
 	}
 
+	/*
+	 * Handle plus separator. Concept is same as tilde, but it one of
+	 * strings ends, it's considered as lower version.
```
Comment (Conan-Kudo) [inline lib/rpmvercmp.c]: Shouldn't this read "it's considered as higher version"?

Comment (ignatenkobrain) [inline lib/rpmvercmp.c]: no.

**Comment by Conan-Kudo** [comment]
Looks good to me! :+1:

**Comment by ignatenkobrain** [comment]
Changed to caret as @ffesti requested in ML.

**Comment by ffesti** [comment]
OK, looks like Fedora is going for a policy that works with the current version compare. If version and release strings are chosen with some care (which they have to be anyway) I do not really see an immediate need for this. Closing for now.

**Comment by ignatenkobrain** [comment]
@ffesti @pmatilai we've discussed this roughly a bit today during our FPC meeting and we pretty much agreed that it would be very good to have this in RPM. Because right now when you want to make post-release snapshot, you would use `1.2+20180101.deadbeef` and then if upstream releases `1.2.1`, you are screwed.

Can we re-open this please?

**Comment by Conan-Kudo** [comment]
@ignatenkobrain That's why you're supposed to do something like `1.2+git20180101.deadbeef`, so that it sorts lower than `1.2.1`.

**Comment by ignatenkobrain** [comment]
I can go further with some worse examples like openssl ;)

`1.1.0+gitdeadbeef > 1.1.0a`

**Comment by Conan-Kudo** [comment]
You crush my soul. :cry:

**Comment by ignatenkobrain** [comment]
I mean if we have tilde, there is a need to have operator which is doing opposite -- that is caret ;)

**Comment by jasontibbitts** [comment]
Hey, this was just pointed out to me, and it seems to perfectly provide a solution to a problem.

As many might be aware, the Fedora packaging committee has periodically taken up the issue of using tilde Version:.  And I've been putting in a load of effort trying to come up with a consistent scheme which actually covers the range of cases that our current scheme does.

One thing I've run into is that tilde is one of a pair of operators you actually need to implement a scheme which consistently places information about snapshots into the Version: field.  If you use tilde, you can (and in some cases must) move prerelease information on snapshots into Version:, but information on post-release snapshots has to stay in Release: because you cannot know what upstream will call the next version.

So you can't do `3.2 < 3.2(snapshot) < 3.2.1` because there's nothing you can fill in for `(snapshot)`.  In my exposition for this I had mentioned a mythical operator '?' or '+' or something which did this.  Turns out this PR provides exactly that, calling it '^'.

I first mentioned this in the context of the work I'm doing in https://pagure.io/packaging-committee/issue/398#comment-540026 and also in my notes for the use of tilde in https://fedoraproject.org/wiki/User:Tibbs/TildeVersioning#Conceptual_issues_with_tilde

@pmatilai also indirectly mentions this in https://pagure.io/packaging-committee/issue/398#comment-540133 and mentions that it can be hacked around by adding random zeroes.  Though in reality, you would have to add multiple zeroes to truly avoid conflicts.

So, while lack of this isn't going to stop the current effort to implement tilde, it would certainly simplify things if we had it.

**Comment by Conan-Kudo** [comment]
@ignatenkobrain I'm not against this, I'm just sad that bad versioning is a thing.

**Comment by hroncok** [comment]
Also, as for the reason this was closed:

> looks like Fedora is going for a policy that works with the current version compare ... I do not really see an immediate need for this

Note that when creating Fedora policies and guidelines, the FPC is forced to make it work with the current version compare (current anything generally), otherwise the policy cannot be used, as it would not work.

If we have this, we can make the policy far simpler. Please reconsider applying this.

**Comment by pmatilai** [comment]
I can't reopen this, perhaps because the original branch was deleted. Anyway, I'm at least willing to discuss and certainly not directly opposed if somebody wants to open a new PR. Not because Fedora this or that but because I think symmetry makes sense: if we have a way force sorting one way, then why not the other way too?

**Pull request closed without merging**